src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java

```diff
@@ -0,0 +1,368 @@
+
+		ResourceFactory<?> resourceFactory = store
+				.getOrComputeIfAbsent( //
+					factoryKey(sharedAnnotation), //
+					__ -> ReflectionSupport.newInstance(sharedAnnotation.factory()), //
+					ResourceFactory.class);
+		ResourceWithLock<?> resource;
+		try {
+			resource = store
+					.getOrComputeIfAbsent( //
+						resourceKey(sharedAnnotation), //
+						__ -> createResource(sharedAnnotation, resourceFactory), //
+						ResourceWithLock.class);
+		}
+		catch (UncheckedParameterResolutionException e) {
+			throw e.getCause();
+		}
+		try {
+			return resource.get();
+		}
+		catch (Exception e) {
+			throw new ParameterResolutionException(
+				"Unable to get the contents of the resource created by `" + sharedAnnotation.factory() + '`', e);
+		}
+	}
+
+	private ResourceWithLock<?> createResource(Shared sharedAnnotation, ResourceFactory<?> resourceFactory) {
+		try {
+			return new ResourceWithLock<>(
+				resourceFactory.create(unmodifiableList(asList(sharedAnnotation.arguments()))));
+		}
+		catch (Exception e) {
+			throw new UncheckedParameterResolutionException(new ParameterResolutionException(
+				"Unable to create a resource from `" + sharedAnnotation.factory() + "`", e));
+		}
+	}
+
+	private void throwIfHasAnnotationWithSameNameButDifferentType(ExtensionContext.Store store,
+			Shared sharedAnnotation) {
+		ResourceFactory<?> presentResourceFactory = //
+			store.getOrDefault(factoryKey(sharedAnnotation), ResourceFactory.class, null);
+
+		if (presentResourceFactory == null) {
+			store.put(keyOfFactoryKey(sharedAnnotation), factoryKey(sharedAnnotation));
+		} else {
+			String presentResourceFactoryName = //
+				store.getOrDefault(keyOfFactoryKey(sharedAnnotation), String.class, null);
+
+			if (factoryKey(sharedAnnotation).equals(presentResourceFactoryName)
+					&& !sharedAnnotation.factory().equals(presentResourceFactory.getClass())) {
+				throw new ParameterResolutionException(
+					"Two or more parameters are annotated with @Shared annotations with the name \""
+							+ sharedAnnotation.name() + "\" but with different factory classes");
+			}
+		}
+	}
+
+	private void throwIfMultipleParametersHaveExactAnnotation(Parameter[] parameters, Shared sharedAnnotation) {
+		long parameterCount = //
+			Arrays.stream(parameters).filter(parameter -> hasAnnotation(parameter, sharedAnnotation)).count();
+		if (parameterCount > 1) {
+			throw new ParameterResolutionException(
+				"A test method has " + parameterCount + " parameters annotated with @Shared with the same "
+						+ "factory type and name; this is redundant, so it is not allowed");
+		}
+	}
+
+	private boolean hasAnnotation(Parameter parameter, Shared sharedAnnotation) {
+		return AnnotationSupport
+				.findAnnotation(parameter, Shared.class)
+				.filter(shared -> shared.factory().equals(sharedAnnotation.factory()))
+				.filter(shared -> shared.name().equals(sharedAnnotation.name()))
+				.isPresent();
+	}
+
+	private long uniqueKey() {
```

Comment:
@nipafx Thanks for renaming this from `newKey()` to `uniqueKey()`, makes more sense. :+1: